consensus/src/main/java/org/apache/iotdb/consensus/multileader/logdispatcher/LogDispatcher.java

```diff
@@ -0,0 +1,284 @@
+          while ((batch = getBatch()).isEmpty()) {
+            // we may block here if there is no requests in the queue
+            bufferedRequest.add(pendingRequest.take());
+            // If write pressure is low, we simply sleep a little to reduce the number of RPC
+            if (pendingRequest.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+              Thread.sleep(MultiLeaderConsensusConfig.MAX_WAITING_TIME_FOR_ACCUMULATE_BATCH_IN_MS);
+            }
+          }
+          // we may block here if the synchronization pipeline is full
+          syncStatus.addNextBatch(batch);
+          // sends batch asynchronously and migrates the retry logic into the callback function
+          sendBatchAsync(batch, new DispatchLogHandler(this, batch));
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        logger.error("Unexpected error in logDispatcher for peer {}", peer, e);
+      }
+      logger.info("{}: Dispatcher for {} exits", impl.getThisNode(), peer);
+    }
+
+    public PendingBatch getBatch() {
+      PendingBatch batch;
+      List<TLogBatch> logBatches = new ArrayList<>();
+      long startIndex = syncStatus.getNextSendingIndex();
+      long maxIndex = impl.getController().getCurrentIndex();
+      long endIndex;
+      if (bufferedRequest.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+        // Use drainTo instead of poll to reduce lock overhead
+        pendingRequest.drainTo(
+            bufferedRequest,
+            MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH - bufferedRequest.size());
+      }
+      if (bufferedRequest.isEmpty()) {
+        // only execute this after a restart
+        endIndex = constructBatchFromWAL(startIndex, maxIndex, logBatches);
+        batch = new PendingBatch(startIndex, endIndex, logBatches);
+        logger.debug("accumulated a {} from wal", batch);
+      } else {
+        Iterator<IndexedConsensusRequest> iterator = bufferedRequest.iterator();
+        IndexedConsensusRequest prev = iterator.next();
+        // Prevents gap between logs. For example, some requests are not written into the queue when
+        // the queue is full. In this case, requests need to be loaded from the WAL
+        constructBatchFromWAL(startIndex, prev.getSearchIndex(), logBatches);
+        constructBatchIndexedFromConsensusRequest(prev, logBatches);
+        endIndex = prev.getSearchIndex();
+        iterator.remove();
+        while (iterator.hasNext()
+            && logBatches.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+          IndexedConsensusRequest current = iterator.next();
+          // Prevents gap between logs. For example, some logs are not written into the queue when
+          // the queue is full. In this case, requests need to be loaded from the WAL
+          if (current.getSearchIndex() != prev.getSearchIndex() + 1) {
```

Comment:
In the scenario current.index == 100, prev.index = 1 and batch size = 10
We construct a batch from wal whose index rang is 2 - 11，and then we call constructBatchIndexedFromConsensusRequest and add the current whose index is 100. I think this will cause an error.

Reply:
This is really a careful bug finding! The processing method is also very simple, we execute a special judgment after the function, if it is full directly return. Since the elements in the bufferedRequest are not deleted at this point, we can continue processing the next time we enter the function.